userland/AppMakefile.mk

```diff
@@ -66,19 +100,19 @@ $$(BUILDDIR)/$(1):
 
 # First step doesn't actually compile, just generate header dependency information
 # More info on our approach here: http://stackoverflow.com/questions/97338
-$$(BUILDDIR)/$(1)/%.o: %.c | $$(BUILDDIR)/$(1)
+$$(BUILDDIR)/$(1)/%.o: %.c | $$(BUILDDIR)/$(1) $$(HDRS)
```

Comment:
That rule says that "in order for the these .c files to convert to these .o files, the directory BUILDDIR/arch must exist and all the files listed in HDRS must exist"

The rules for creating any files that may appear in HDRS come from elsewhere (in the case of libnrfserialization they come from the included `Makefile.app`, as do the additions the HDRS variable itself)

Upon a second look it probably should be
`$$(BUILDDIR)/$(1)/%.o: %.c $$(HDRS) | $$(BUILDDIR)/$(1)`
instead of an order-only rule. Will fix that momentarily

but dependencies are never listed in the build rules.. they're dependencies. You can't compile those .c files unless the headers they #include exist. I'm confused what's confusing you.